### PR TITLE
Several changes

### PR DESCRIPTION
**Pull request by ryanenzo**
- Add ability to generate both .h and .cpp file in one run of kdwsdl2cpp 
- Add support for namespace-mapping to control class-name generation 
  (e.g. to generate "MYPREFIX__MyClass" instead of "TNS__MyClass")
- Add endpoint() and soapVersion() accessors in generated Wsdl Service classes
- Fixes to prevent generation of some invalid C++

### CODE REVIEW

File: kdwsdl2cpp/libkode/style.cpp [hunk 1]
```diff
@@ -83,6 +83,9 @@ QString Style::lowerFirst( const QString &str )
 
 QString Style::makeIdentifier( const QString &str )
 {
+  if ( str.isEmpty() )
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/libkode/style.cpp]: Strange place for this check, "" isn't a valid identifier in the first place... No big deal but it makes me wonder, who's calling this with an empty identifier?

File: kdwsdl2cpp/src/compiler.cpp [hunk 2]
```diff
@@ -103,7 +103,6 @@ void Compiler::parse(const QDomElement &element)
 
         definitions.fixUpDefinitions(/*&context, element*/);
 
-        KODE::Code::setDefaultIndentation(4);
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/compiler.cpp]: Why was this removed?

File: kdwsdl2cpp/src/converter_complextype.cpp [hunk 3]
```diff
@@ -643,6 +644,7 @@ void Converter::createComplexTypeSerializer(KODE::Class &newClass, const XSD::Co
         bool first = true;
         Q_FOREACH (const XSD::Attribute &attribute, attributes) {
             const QString attrName = attribute.name();
+            if (attrName.isEmpty()) { continue; }
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/converter_complextype.cpp]: On 3 lines please.

File: kdwsdl2cpp/src/creator.h [hunk 4]
```diff
@@ -21,16 +21,30 @@
...
 #include <libkode/class.h>
+#include <libkode/file.h>
+#include <libkode/license.h>
+#include <libkode/printer.h>
 
 namespace KWSDL
 {
 
-class Creator
+class Creator : private KODE::Printer
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/creator.h]: I would find it more readable to have the printer as a member variable.

File: kdwsdl2cpp/src/main.cpp [hunk 5]
```diff
@@ -123,6 +147,40 @@ int main(int argc, char **argv)
...
+            if (!argv[arg]) {
+                showHelp(argv[0]);
+                return 1;
+            }
+            QString mapping = argv[arg];
+            if (mapping.startsWith('@')) {
+                QString mappingFileName = mapping.mid(1);
+                QFile file(mappingFileName);
+                if (!file.open(QIODevice::ReadOnly)) {
+                    fprintf(stderr, "Error reading %s: %s\n", mappingFileName.toLatin1().data(), file.errorString().toLatin1().data());
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/main.cpp]: `QFile::encodeName(mappingFileName).constData()` would be more correct (in case of non-latin1 chars in the path)

File: kdwsdl2cpp/src/main.cpp [hunk 6]
```diff
@@ -123,6 +147,40 @@ int main(int argc, char **argv)
...
+                    showHelp(argv[0]);
+                    return 1;
+                }
+
+                while(!file.atEnd()) {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/main.cpp]: missing space after `while`

File: kdwsdl2cpp/src/main.cpp [hunk 7]
```diff
@@ -123,6 +147,40 @@ int main(int argc, char **argv)
...
+                    QString mapping = file.readLine().trimmed();
+                    if (mapping.startsWith("#")) { continue; }
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/main.cpp]: `'#'` and indent on 3 lines

File: kdwsdl2cpp/src/main.cpp [hunk 8]
```diff
@@ -163,19 +221,52 @@ int main(int argc, char **argv)
...
+    }
+
+
+    if (both) {
+        Settings::self()->setGenerateHeader(true);
+        Settings::self()->setGenerateImplementation(true);
+        Settings::self()->setHeaderFileName(baseFile + ".h");
+        Settings::self()->setImplementationFileName(baseFile + ".cpp");
+    }
+    else if (impl) {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/main.cpp]: join with previous line

File: kdwsdl2cpp/src/main.cpp [hunk 9]
```diff
@@ -163,19 +221,52 @@ int main(int argc, char **argv)
...
+        Settings::self()->setGenerateHeader(false);
+        Settings::self()->setGenerateImplementation(true);
+        Settings::self()->setHeaderFileName(headerFile);
+        Settings::self()->setImplementationFileName(outputFile.fileName());
+    } 
+    else {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/main.cpp]: join with previous line

File: kdwsdl2cpp/src/main.cpp [hunk 10]
```diff
@@ -163,19 +221,52 @@ int main(int argc, char **argv)
...
     Settings::self()->setExportDeclaration(exportMacro);
     Settings::self()->setNameSpace(nameSpace);
+    Settings::self()->setNamespaceMapping(nsmapping);
     Settings::self()->setOptionalElementType(optionalElementType);
     Settings::self()->setKeepUnusedTypes(keepUnusedTypes);
     Settings::self()->setImportPathList(importPathList);
     Settings::self()->setUseLocalFilesOnly(useLocalFilesOnly);
     Settings::self()->setHelpOnMissing(helpOnMissing);
+
+    //fprintf(stderr, "Output dir: %s\n", Settings::self()->outputDirectory().toLatin1().data());
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/main.cpp]: I think you can remove all this debugging now

File: kdwsdl2cpp/src/settings.cpp [hunk 11]
```diff
@@ -101,11 +102,17 @@ QString Settings::wsdlFileName() const
...
-    mOutputFileName = outputFileName;
+    mHeaderFileName = headerFileName;
 }
 
+void Settings::setImplementationFileName(const QString &implementationFileName)
+{
+    mImplementationFileName = implementationFileName;
+}
+
+/*
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/settings.cpp]: kill dead code

File: src/KDSoapClient/KDSoapClientInterface.cpp [hunk 12]
```diff
@@ -58,7 +58,7 @@ void KDSoapClientInterface::setSoapVersion(KDSoapClientInterface::SoapVersion ve
     d->m_version = static_cast<KDSoap::SoapVersion>(version);
 }
 
-KDSoapClientInterface::SoapVersion KDSoapClientInterface::soapVersion()
+KDSoapClientInterface::SoapVersion KDSoapClientInterface::soapVersion() const
```
Comment (dfaure-kdab) [inline src/KDSoapClient/KDSoapClientInterface.cpp]: This is unrelated, and I'm pretty sure I saw it in a different review request already, please remove from this commit.

File: kdwsdl2cpp/src/main.cpp [hunk 13]
```diff
@@ -163,19 +221,52 @@ int main(int argc, char **argv)
...
+        Settings::self()->setGenerateHeader(false);
+        Settings::self()->setGenerateImplementation(true);
+        Settings::self()->setHeaderFileName(headerFile);
+        Settings::self()->setImplementationFileName(outputFile.fileName());
+    } 
+    else {
+        Settings::self()->setGenerateHeader(true);
+        Settings::self()->setGenerateImplementation(false);
+        Settings::self()->setHeaderFileName(outputFile.fileName());
+        Settings::self()->setImplementationFileName("UNUSED");
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/main.cpp]: Why call it then? To see an obvious error if this is ever used? ;-)

File: kdwsdl2cpp/src/main.cpp [hunk 14]
```diff
@@ -163,19 +221,52 @@ int main(int argc, char **argv)
         return 1;
     }
 
+    // if you're saying "just make the impl-file", you can't
+    //    also say "make both the header and the impl"
+    if (both && impl) {
```
Comment (dfaure-kdab) [inline kdwsdl2cpp/src/main.cpp]: Same if specifying -o and -both, right? This should give an error too.
To do it here you'd need another bool, the alternative would be to test it when parsing -both (outputFile should still be empty) and when parsing -o (both should be false).

File: kdwsdl2cpp/libkode/style.cpp (hunk 1 above)
Comment (ryanenzo) [reply kdwsdl2cpp/libkode/style.cpp]: I don't recall the situation, but it was causing a crash. Thus this was added.

File: kdwsdl2cpp/src/compiler.cpp (hunk 2 above)
Comment (ryanenzo) [reply kdwsdl2cpp/src/compiler.cpp]: Not removed, just moved into the Creator class.

File: kdwsdl2cpp/libkode/style.cpp (hunk 1 above)
Comment (dfaure-kdab) [reply kdwsdl2cpp/libkode/style.cpp]: Yes the fact that this would crash is obvious, given that at(0) in the implementation.

But it's still an invalid use of this method, in my opinion.
Make it a `Q_ASSERT(!str.isEmpty())` for clarity that this is the contract, but then you'll still have to fix the caller.

Comment (dfaure-kdab) [reply kdwsdl2cpp/libkode/style.cpp]: This is still pending.

Comment (ryanenzo) [reply kdwsdl2cpp/libkode/style.cpp]: Shoudl be done now.
